dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala

```diff
@@ -126,18 +128,36 @@ object DistriOptimizer {
           val cached = modelIter.next()
           val syWStart = System.nanoTime()
           val weightsResult = parameters.getWeights(cached.modelWeights.head)
-          val tensorBuffer = new Array[(Tensor[T], Tensor[T])](_subModelNumber)
+          val batchBuffer = new Array[(Activity, Activity)](_subModelNumber)
           tasks += Engine.default.invoke(() => {
             val batch = data.next()
             var b = 0
-            require(batch.data.size(1) == batch.labels.size(1),
-              "data and label batch size not match")
-            require(batch.data.size(1) >= _subModelNumber,
-              "total batch size should be divided by total core number")
-            val stackSize = batch.data.size(1) / _subModelNumber
+            val tensorBatchType = batch.data match {
+              case tensor: Tensor[T] =>
+                require(batch.data.toTensor[T].size(1) == batch.labels.toTensor[T].size(1),
+                  "data and label batch size not match")
+                require(batch.data.toTensor[T].size(1) >= _subModelNumber,
+                  "total batch size should be divided by total core number")
+                true
+              case table: Table =>
+                require(batch.data.toTable.length == batch.labels.toTable.length,
+                  "data and label batch size not match")
+                require(batch.data.toTable.length >= _subModelNumber,
+                  "total batch size should be divided by total core number")
+                false
+            }
+            val stackSize = tensorBatchType match {
+              case true => batch.data.toTensor[T].size(1) / _subModelNumber
+              case false => batch.data.toTable.length / _subModelNumber
+            }
             while (b < _subModelNumber) {
-              tensorBuffer(b) = (batch.data.narrow(1, b * stackSize + 1, stackSize),
-                batch.labels.narrow(1, b * stackSize + 1, stackSize))
+              batchBuffer(b) = tensorBatchType match {
+                case true => (batch.data.toTensor[T].narrow (1, b * stackSize + 1, stackSize),
+                  batch.labels.toTensor[T].narrow (1, b * stackSize + 1, stackSize))
+                case false =>
+                  (NarrowTable(b * stackSize + 1, stackSize).updateOutput(batch.data.toTable),
```

Comment:
It is better to create a NarrowTable out of the while loop. In that case, the narrowTable will not be created each time.
Like that:
define `val narrowTable = NarrowTable(b * stackSize + 1, stackSize)` in line 153, 
and call `narrowTable.updateOutput(batch.data.toTable)` here